File: leetcode/14_longest_common_prefix.cpp
/*
 * @lc app=leetcode id=14 lang=cpp
 *
 * [14] Longest Common Prefix
 */

#include <string>
#include <vector>

// @lc code=start
class Solution {
 public:
  std::string longestCommonPrefix(std::vector<std::string>& strs) {
    std::string prefix = strs[0];
    for (int i = 1; i < strs.size(); i++) {
      while (strs[i].find(prefix) != 0) {
        prefix = prefix.substr(0, prefix.length() - 1);
        if (prefix.empty()) return "";
      }
    }
    return prefix;
  }
};
// @lc code=end

File: leetcode/14_longest_common_prefix.rs
/*
 * @lc app=leetcode id=14 lang=rust
 *
 * [14] Longest Common Prefix
 */

// @lc code=start
impl Solution {
    pub fn longest_common_prefix(strs: Vec<String>) -> String {
        let mut prefix = strs[0].clone();
        for (i, s) in strs[1..].iter().enumerate() {
            while !s.starts_with(&prefix) {
                prefix.pop();
                if prefix.is_empty() {
                    return String::new();
                }
            }
        }

        prefix
    }
}
// @lc code=end
